src/dune_rules/lib.ml

```diff
@@ -1198,7 +1199,7 @@ end = struct
           in
           Some (Select { select with choices }))
 
-  let resolve_complex_deps db deps ~private_deps : resolved_deps Memo.t =
+  let resolve_complex_deps ?modes db deps ~private_deps : resolved_deps Memo.t =
```

Comment:
modes should not be optional but rather passed every time and set to the mode of the library we initialize